Reading data from LabelledData.txt and printing the header to visualize the data.
Given input(Training Data) is having 1st and 2nd column as NAN and also there are 1387 Total Records.

In [194]:
import pandas as pd
data = pd.read_csv("LabelledData.txt", sep=',',error_bad_lines=False,header=None)
print (data.head(5))
print (data.shape)

                                                   0   1   2         3
0  how did serfdom develop in and then leave russ... NaN NaN   unknown
1  what films featured the character popeye doyle ?  NaN NaN      what
2  how can i find a list of celebrities ' real na... NaN NaN   unknown
3  what fowl grabs the spotlight after the chines... NaN NaN      what
4                   what is the full form of .com ?  NaN NaN      what
(1387, 4)


b'Skipping line 136: expected 4 fields, saw 6\nSkipping line 143: expected 4 fields, saw 5\nSkipping line 159: expected 4 fields, saw 5\nSkipping line 168: expected 4 fields, saw 5\nSkipping line 182: expected 4 fields, saw 5\nSkipping line 186: expected 4 fields, saw 6\nSkipping line 235: expected 4 fields, saw 5\nSkipping line 246: expected 4 fields, saw 7\nSkipping line 250: expected 4 fields, saw 5\nSkipping line 254: expected 4 fields, saw 5\nSkipping line 258: expected 4 fields, saw 5\nSkipping line 260: expected 4 fields, saw 6\nSkipping line 271: expected 4 fields, saw 6\nSkipping line 298: expected 4 fields, saw 5\nSkipping line 326: expected 4 fields, saw 5\nSkipping line 356: expected 4 fields, saw 5\nSkipping line 379: expected 4 fields, saw 5\nSkipping line 414: expected 4 fields, saw 6\nSkipping line 420: expected 4 fields, saw 6\nSkipping line 457: expected 4 fields, saw 5\nSkipping line 476: expected 4 fields, saw 6\nSkipping line 484: expected 4 fields, saw 5\nSkipping

In [195]:
data = data.dropna(axis=1, how='all')
data.columns = ["question", "label"]
data['label'] = data['label'].str.strip()
print (data.head(5))
print (data.shape)

                                            question    label
0  how did serfdom develop in and then leave russ...  unknown
1  what films featured the character popeye doyle ?      what
2  how can i find a list of celebrities ' real na...  unknown
3  what fowl grabs the spotlight after the chines...     what
4                   what is the full form of .com ?      what
(1387, 2)


After Dropping NAN values and rechecking the size of the dataframe

In [196]:
data['label'] = data['label'].astype(str) 
print (data['label'].unique())
data.label.value_counts()

['unknown' 'what' 'when' 'who' 'affirmation']


what           574
who            371
unknown        250
affirmation    103
when            89
Name: label, dtype: int64

Trying to get the insights about the training data by finding out the unique target labels and its coressponding count.

In [197]:
data[data['label']== 'unknown']

question    label
0     how did serfdom develop in and then leave russ...  unknown
2     how can i find a list of celebrities ' real na...  unknown
9     name the scar-faced bounty hunter of the old w...  unknown
11     why do heavier objects travel downhill faster ?   unknown
17    how many jews were executed in concentration c...  unknown
22                            name 11 famous martyrs .   unknown
28                name a golf course in myrtle beach .   unknown
29           which two states enclose chesapeake bay ?   unknown
32    how many points make up a perfect fivepin bowl...  unknown
33    which company that manufactures video-game har...  unknown
34    how many community chest cards are there in mo...  unknown
38    where do the adventures of `` the swiss family...  unknown
40    how can i register my website in yahoo for fre...  unknown
42         how many people in the world speak french ?   unknown
43    how many inches over six feet is the venus de ...  unknown
44    how do they find or choose witnesses to an exe...  unknown
48     how do you ask a total stranger out on a date ?   unknown
53           where did the term `` 86ed '' come from ?   unknown
60                                 where is the loop ?   unknown
65    which city has the oldest relationship as a si...  unknown
67                        why do people get calluses ?   unknown
75    how do i make fuel bricks from recycled newspa...  unknown
78    how do i log on to home page at headquarters u...  unknown
81    how many species of the great white shark are ...  unknown
85    how many villi are found in the small intestin...  unknown
89    in which year was new zealand excluded from th...  unknown
104        which of the following was rhodes scholar ?   unknown
106                      how do you make a paintball ?   unknown
109              where is the highest point in japan ?   unknown
110                    where do chihuahuas come from ?   unknown
...                                                 ...      ...
826   why are electric cars less efficient in the no...  unknown
830   name the various costumed personas of dr. henr...  unknown
848      which soft drink does madonna advertise for ?   unknown
851   how many double-word-score spaces are there on...  unknown
854   why were the american soldiers in wwi called `...  unknown
856               how many cds has garth brooks sold ?   unknown
857   name the vessel used by the atari force in the...  unknown
858   where are the leading medical groups specializ...  unknown
862              how many gutenberg bibles are there ?   unknown
863        name the largest country in south america .   unknown
873   which condiment was once sold as a patent medi...  unknown
875   why did the world enter a global depression in...  unknown
891   how many home runs did lou gehrig have during ...  unknown
893      where can i find information on george bush ?   unknown
897   how many boys play the game in winslow homer '...  unknown
905   how many astronauts manned each project mercur...  unknown
909   name the only extant trilogy of classical gree...  unknown
910                           where is logan airport ?   unknown
917   which of the following men was not married to ...  unknown
920                how do you find oxidation numbers ?   unknown
923   how many students attend the university of mas...  unknown
930   how do you look up criminal records on the int...  unknown
950       whose video is titled shape up with arnold ?   unknown
985                          whose cupboard was bare ?   unknown
1011  whom did lauren bacall marry after her husband...  unknown
1063  whose image is alleged to be on the shroud of ...  unknown
1098  whose kidnaping was termed the crime of the ce...  unknown
1190  whom did friz freleng add to the warner bros. ...  unknown
1219  whose biography by maurice zolotow is titled s...  unknown
1269  whose acceptance speech of more than 3 minutes...  unknown

[250 rows x 2 co

In [198]:
data.shape

(1387, 2)

As it can be seen from the initial analysis out of these 250 questions most of them can be assigned the labels which we explored above therefore trying to assign the label to these 250 questions and will also try to normalize ' who' and '  who' which are currently being treated different because of the fomrating issue.

In [199]:
KeyWords = ["how","where","which","whom","whose","why","who"]
temp_dict =   {}
l1 = []
def Unknown_label(data):
    return {row['question']:row['label'] for index, row in data.iterrows() if row['label'] == 'unknown'}
def label_assign(result):
    for k,v in result.items():
        first_word = k.split(" ")[0]
        if first_word in KeyWords:
            new_word = k.split(" ")[0]
            temp_dict.update({k:new_word})  
        else:
            temp_dict.update({k:'unknown'})
    return temp_dict
result = Unknown_label(data)
UnknownValues= label_assign(result)

In [200]:
data.shape

(1387, 2)

Building and assigning the questions and labels to the dataframe that we got from the new build dictionary

In [201]:
s = pd.Series(UnknownValues)
data_new = pd.DataFrame({'question':s.index, 'label':s.values})
data_new = data_new.reindex_axis(['question','label'], axis=1)
data_new.shape

C:\Users\jeetk\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: '.reindex_axis' is deprecated and will be removed in a future version. Use '.reindex' instead.
  This is separate from the ipykernel package so we can avoid doing imports until


(249, 2)

In [202]:
print (data_new['label'].unique())
data_new.label.value_counts()

['unknown' 'how' 'where' 'which' 'who' 'whom' 'whose' 'why']


how        118
where       53
unknown     29
which       19
why         17
whose        9
whom         3
who          1
Name: label, dtype: int64

In [203]:
data1 = data
data1.drop(data1[data1.label == 'unknown' ].index, inplace=True)
bigdata = data.append(data_new, ignore_index=True)
print (bigdata['label'].unique())
bigdata.label.value_counts()


['what' 'when' 'who' 'affirmation' 'unknown' 'how' 'where' 'which' 'whom'
 'whose' 'why']


what           574
who            372
how            118
affirmation    103
when            89
where           53
unknown         29
which           19
why             17
whose            9
whom             3
Name: label, dtype: int64

In [204]:
bigdata.to_csv('LabelledData_Clean.txt', sep=',', encoding='utf-8')

In [205]:
import string
import pandas as pd
from nltk import word_tokenize
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer
import time
from sklearn.metrics import classification_report
from sklearn import  svm


header = ["question", "label"]
data = pd.read_csv('LabelledData_Clean.txt',sep=',',header=None, names=header)

train_data = data[:1300]
test_data = data[1301:]

def remove_non_ascii(textVal):
    return ''.join([i if ord(i) < 128 else ' ' for i in textVal])

def tokenize(text):
    text = "".join([ch for ch in text if ch not in string.punctuation])
    tokens = word_tokenize(text)
    stems = stem_tokens(tokens, stemmer)
    return stems

stemmer = PorterStemmer()
def stem_tokens(tokens, stemmer):
    stemmed = []
    for item in tokens:
        stemmed.append(stemmer.stem(item))
    return stemmed

vectorizer = TfidfVectorizer(tokenizer=tokenize, stop_words='english')
train_vectors = vectorizer.fit_transform(train_data.question)
test_vectors = vectorizer.transform(test_data.question)
classifier_linear = svm.SVC(kernel='linear')
classifier_linear.fit(train_vectors, train_data.label)
prediction_linear = classifier_linear.predict(test_vectors)

In [206]:
import pandas as pd

#new_data = pd.read_csv("LabelledData_Clean.txt",error_bad_lines=False)

header = ["label", "question"]
new_data = pd.read_csv("Test.csv",sep=':',header=None, names=header)

test_data_question = new_data["question"]

vectored =  vectorizer.transform(new_data.question)
predicted_label = classifier_linear.predict(vectored)

new_dict = {
"question": test_data_question,     
"Predicted" : predicted_label   
 }
new_dict_df = pd.DataFrame(new_dict)
new_dict_df.to_csv("Results.csv", sep=',')